packages/flutter/lib/src/rendering/object.dart

```diff
@@ -57,190 +59,215 @@ class PaintingCanvas extends Canvas {
 /// not hold a reference to the canvas across operations that might paint
 /// child render objects.
 class PaintingContext {
-  /// Construct a painting context at a given offset with the given bounds
-  PaintingContext.withOffset(Offset offset, Rect paintBounds) {
-    _containerLayer = new ContainerLayer(offset: offset);
-    _startRecording(paintBounds);
+  PaintingContext._(this._containerLayer, this._paintBounds);
+
+  final ContainerLayer _containerLayer;
+  final Rect _paintBounds;
+
+  /// Repaint the given render object.
+  ///
+  /// The render object must have a composited layer and must be in need of
+  /// painting. The render object's layer is re-used, along with any layers in
+  /// the subtree that don't need to be repainted.
+  static void repaint(RenderObject root) {
+    assert(root.hasLayer);
+    assert(root.needsPaint);
+    _repaintCompositedChild(root);
   }
 
-  /// Construct a painting context for painting into the given layer with the given bounds
-  PaintingContext.withLayer(ContainerLayer containerLayer, Rect paintBounds) {
-    _containerLayer = containerLayer;
-    _startRecording(paintBounds);
+  /// Paint a child render object.
+  ///
+  /// If the child has its own composited layer, the child will be composited
+  /// into the layer subtree associated with this painting context. Otherwise,
+  /// the child will be painted into the current PictureLayer for this context.
+  void paintChild(RenderObject child, Offset offset) {
+    if (child.hasLayer) {
+      _stopRecordingIfNeeded();
+      _compositeChild(child, offset);
+    } else {
+      child._paintWithContext(this, offset);
+    }
   }
 
-  /// A backdoor for testing that lets the test set a specific canvas
-  PaintingContext.forTesting(this._canvas);
+  void _compositeChild(RenderObject child, Offset offset) {
+    assert(!_isRecording);
+    assert(child.hasLayer);
 
-  ContainerLayer _containerLayer;
-  /// The layer contain all the composting layers that will be used for this context
-  ContainerLayer get containerLayer => _containerLayer;
+    // Create a layer for our child, and paint the child into it.
+    if (child.needsPaint) {
+      _repaintCompositedChild(child);
+    } else {
+      assert(child._layer != null);
+      child._layer.detach();
+    }
+    _appendLayer(child._layer, offset);
+  }
+
+  void _appendLayer(Layer layer, Offset offset) {
+    assert(!_isRecording);
+    layer.offset = offset;
+    _containerLayer.append(layer);
+  }
 
+  static void _repaintCompositedChild(RenderObject child) {
+    child._layer ??= new ContainerLayer();
+    child._layer.removeAllChildren();
+    PaintingContext childContext = new PaintingContext._(child._layer, child.paintBounds);
+    child._paintWithContext(childContext, Offset.zero);
+    childContext._stopRecordingIfNeeded();
+  }
+
+  bool get _isRecording {
+    final bool hasCanvas = (_canvas != null);
+    assert(() {
+      if (hasCanvas) {
+        assert(_currentLayer != null);
```

Comment:
Yeah, I thought about doing something like that, but the trouble is when the assert fires, it's harder to know what's going wrong.  This way the line number tells you whether there is a canvas and which variable is messed up.